### PR TITLE
feat: add metadata_queries api layer

### PR DESCRIPTION
**Pull request by manishPh**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @manishPh has signed the CLA. Thanks for the pull request!

File: src/api/MetadataQuery.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    errorCallback: ElementsErrorCallback;
+
+    /**
+     * Creates a key for the metadata cache
+     *
+     * @param {string} id - metadata template
+     * @return {string} key
+     */
+    getCacheKey(id: string): string {
+        return `${CACHE_PREFIX_METADATA}query_${id}`;
```
Comment (ConradJChan) [inline src/api/MetadataQuery.js]: Should we just keep them separate and create a new constant for the metadata query cache prefix?

File: src/api/MetadataQuery.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    /**
+     * API for querying enterprise metadata
+     * @param {Object} query - metadata query object
+     * @param {Function} successCallback - Function to call with results
+     * @param {Function} errorCallback - Function to call with errors
+     * @param {boolean|void} [options.forceFetch] - Bypasses the cache
+     * @return {void}
+     */
+    queryMetadata(
+        query: Object,
```
Comment (jstoffan) [inline src/api/MetadataQuery.js]: Can these parameters be more strongly-typed?

File: src/api/MetadataQuery.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    };
+
+    /**
+     * Does the network request to metadata query API
+     * @param {Object} query query object with SQL Clauses like properties
+     * @return {Promise}
+     */
+    queryMetadataRequest(query: Object): Promise<void> {
+        if (this.isDestroyed()) {
+            return Promise.reject();
```
Comment (jstoffan) [inline src/api/MetadataQuery.js]: I think we either `return` or call `successCallback` in most other API modules. What error would be shown to the user in this case?

File: src/api/MetadataQuery.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+     *
+     * @return {boolean} if query results are loaded
+     */
+    isLoaded(): boolean {
+        const cache: APICache = this.getCache();
+        return cache.has(this.key);
+    }
+
+    /**
+     * Returns the results
```
Comment (jstoffan) [inline src/api/MetadataQuery.js]: It looks like the method doesn't currently return anything.

File: src/api/MetadataQuery.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        cache.set(this.key, data);
+        this.finish();
+    };
+
+    /**
+     * Does the network request to metadata query API
+     * @param {Object} query query object with SQL Clauses like properties
+     * @return {Promise}
+     */
+    queryMetadataRequest(query: Object): Promise<void> {
```
Comment (jstoffan) [inline src/api/MetadataQuery.js]: Can both this and the success handler be grouped more closely with its caller, `queryMetadata`?

File: src/api/MetadataQuery.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        successCallback: Function,
+        errorCallback: ElementsErrorCallback,
+        options: Object = {},
+    ): void {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        const { context = {} } = options;
+        this.key = this.getCacheKey(context.id);
```
Comment (ConradJChan) [inline src/api/MetadataQuery.js]: What is the `context` and `id` represent here? Does it make sense to do something similar to `Search.js` where the key is the encoded query?

File: src/api/MetadataQuery.js (hunk 3 above)
Comment (manishPh) [reply src/api/MetadataQuery.js]: Both `Folder.js` and `Recents.js` we `return Promise.reject();`. Other files, we simply `return`

File: src/api/MetadataQuery.js (hunk 6 above)
Comment (manishPh) [reply src/api/MetadataQuery.js]: The idea is `context` object will have some useful info about the query thats hard to extract from query object. e.g.enterprise_id, template_key, ancestor_folder_id etc. This is different from the `Search.js` where the query is a simple `string`. 
Component using this API should do the extraction and pass down this info for context. `id` could be generated as a unique cache property for each query.

File: flow-typed/box-ui-elements.js [hunk 7]
```diff
@@ -892,3 +892,35 @@ type AdditionalVersionInfo = {
     currentVersionId?: ?string,
     updateVersionToCurrent: () => void,
 };
+
+type MetadataQeuryResponseEntryEnterprise = {
```
Comment (jstoffan) [inline flow-typed/box-ui-elements.js]: We're trying to eliminate global types. Can these types live in something like `src/common/types/metadataQueries.js`?

Comment (manishPh) [reply flow-typed/box-ui-elements.js]: Will make that change.

**Pull request merged**